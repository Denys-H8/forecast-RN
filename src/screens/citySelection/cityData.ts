export const cities = [
  {
    id: 706483,
    name: 'Kharkiv',
    state: '',
    country: 'UA',
    coord: {
      lon: 36.25,
      lat: 50.0,
    },
  },
  {
    id: 703448,
    name: 'Kyiv',
    state: '',
    country: 'UA',
    coord: {
      lon: 30.516666,
      lat: 50.433334,
    },
  },
  {
    id: 703845,
    name: 'Kryvyi Rih',
    state: '',
    country: 'UA',
    coord: {
      lon: 33.349998,
      lat: 47.916668,
    },
  },
  {
    id: 698740,
    name: 'Odessa',
    state: '',
    country: 'UA',
    coord: {
      lon: 30.732622,
      lat: 46.477474,
    },
  },
  {
    id: 702550,
    name: 'Lviv',
    state: '',
    country: 'UA',
    coord: {
      lon: 24.023239,
      lat: 49.838261,
    },
  },
  {
    id: 707471,
    name: 'Ivano-Frankivsk',
    state: '',
    country: 'UA',
    coord: {
      lon: 24.709721,
      lat: 48.921501,
    },
  },
  {
    id: 689558,
    name: 'Vinnytsia',
    state: '',
    country: 'UA',
    coord: {
      lon: 28.48097,
      lat: 49.23278,
    },
  },
  {
    id: 709930,
    name: 'Dnipro',
    state: '',
    country: 'UA',
    coord: {
      lon: 34.98333,
      lat: 48.450001,
    },
  },
  {
    id: 687700,
    name: 'Zaporizhia',
    state: '',
    country: 'UA',
    coord: {
      lon: 35.183331,
      lat: 47.816669,
    },
  },
  {
    id: 700569,
    name: 'Mykolayiv',
    state: '',
    country: 'UA',
    coord: {
      lon: 31.9974,
      lat: 46.965912,
    },
  },
  {
    id: 693805,
    name: 'Simferopol',
    state: '',
    country: 'UA',
    coord: {
      lon: 34.11079,
      lat: 44.957191,
    },
  },
  {
    id: 694423,
    name: 'Sevastopol',
    state: '',
    country: 'UA',
    coord: {
      lon: 33.533333,
      lat: 44.599998,
    },
  },
  {
    id: 709717,
    name: 'Donetsk',
    state: '',
    country: 'UA',
    coord: {
      lon: 37.799999,
      lat: 48.0,
    },
  },
  {
    id: 696643,
    name: 'Poltava',
    state: '',
    country: 'UA',
    coord: {
      lon: 34.54073,
      lat: 49.593731,
    },
  },
  {
    id: 706448,
    name: 'Kherson',
    state: '',
    country: 'UA',
    coord: {
      lon: 32.617802,
      lat: 46.655811,
    },
  },
  {
    id: 704147,
    name: 'Kremenchuk',
    state: '',
    country: 'UA',
    coord: {
      lon: 33.41972,
      lat: 49.097252,
    },
  },
  {
    id: 686967,
    name: 'Zhytomyr',
    state: '',
    country: 'UA',
    coord: {
      lon: 28.676689,
      lat: 50.26487,
    },
  },
  {
    id: 702569,
    name: 'Lutsk',
    state: '',
    country: 'UA',
    coord: {
      lon: 25.34244,
      lat: 50.759319,
    },
  },
  {
    id: 701404,
    name: 'Melitopol',
    state: '',
    country: 'UA',
    coord: {
      lon: 35.373055,
      lat: 46.84222,
    },
  },
  {
    id: 710791,
    name: 'Cherkasy',
    state: '',
    country: 'UA',
    coord: {
      lon: 32.062069,
      lat: 49.428539,
    },
  },
];
