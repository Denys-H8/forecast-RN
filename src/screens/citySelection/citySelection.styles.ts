import { StyleSheet } from 'react-native';
import constants from 'src/constants';

export const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: constants.BG_COLOR,
  },
  loadingContainer: {
    flex: 1,
    backgroundColor: constants.BG_COLOR,
    alignItems: 'center',
    justifyContent: 'center',
  },
  separator: {
    height: 2,
    backgroundColor: constants.SUB_COLOR,
  },
});
