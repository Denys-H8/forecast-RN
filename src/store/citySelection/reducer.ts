import * as ACTION from './actionTypes';

type Props = {
  city: TCity;
  usedCities: TCity[];
  query: string;
  data: TData[]; // searched cities and their id
  weatherData: TGroupWeather | null;
  isLoading: boolean;
};

export const initialState: Props = {
  city: {
    id: 706483,
    name: 'Kharkiv',
    lat: 49.988358,
    lon: 36.232845,
  },
  usedCities: [
    {
      id: 706483,
      name: 'Kharkiv',
      lon: 36.25,
      lat: 50.0,
    },
  ],
  query: '',
  data: [],
  weatherData: null,
  isLoading: false,
};

export const cityReducer = (state = initialState, action: { type: string; payload?: any }) => {
  switch (action.type) {
    case ACTION.QUERY_CHANGED:
      return {
        ...state,
        query: action.payload.query,
        data: action.payload.data,
      };
    case ACTION.GET_WEATHER_FOR_CITIES_SUCESS:
      return {
        ...state,
        weatherData: action.payload,
      };
    case ACTION.RESET_QUERY:
      return {
        ...state,
        query: '',
      };
    case ACTION.CHANGE_CITY: {
      let usedCitiesNew = state.usedCities;

      if (state.usedCities.filter((e) => e.id === action.payload.id).length === 0) {
        usedCitiesNew.push(action.payload);
      }

      return {
        ...state,
        city: action.payload,
        usedCities: usedCitiesNew,
      };
    }
    case ACTION.SET_LOADING:
      return {
        ...state,
        isLoading: action.payload,
      };
    default:
      return state;
  }
};

export type TGroupWeather = {
  list: TGroupWeatherItem[];
  city: {
    id: number;
    name: string;
    coord: {
      lat: string;
      lon: string;
    };
    sunrise: number;
    sunset: number;
  };
};

export type TData = {
  id: number;
  name: string;
  state: string;
  country: string;
  coord: {
    lon: number;
    lat: number;
  };
};

export type TCity = {
  id: number;
  name: string;
  lat: number;
  lon: number;
};

export type TGroupWeatherItem = {
  coord: {
    lon: number;
    lat: number;
  };
  sys: {
    sunrise: number;
    sunset: number;
  };
  weather: {
    description: string;
  }[];
  main: {
    temp: number;
    feels_like: number;
    temp_min: number;
    temp_max: number;
  };
  visibility: number;
  wind: {
    speed: number;
    deg: number;
  };
  clouds: {
    all: number;
  };
  dt: number;
  id: number;
  name: string;
};
