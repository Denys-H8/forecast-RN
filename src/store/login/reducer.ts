import * as ACTION from './actionTypes';

export type TState = {
  user: TUser | null;
  loadingLogin: boolean;
  email: string;
  password: string;
  btnDisabled: boolean;
  emailError: boolean;
  passwordError: boolean;
};

export const initialState: TState = {
  user: null,
  loadingLogin: true,
  email: '',
  password: '',
  btnDisabled: true,
  emailError: false,
  passwordError: false,
};

export const loginReducer = (state = initialState, action: { type: string; payload?: any }) => {
  switch (action.type) {
    case ACTION.LOGIN_STARTED:
      return state;
    case ACTION.LOGIN_SUCCESS:
      return {
        ...state,
        user: action.payload,
        email: '',
        password: '',
      };
    case ACTION.LOGOUT_SUCCESS:
      return {
        ...state,
        user: null,
      };
    case ACTION.CHANGED_EMAIL:
      return {
        ...state,
        email: action.payload,
        emailError: false,
      };
    case ACTION.CHANGED_PASSWORD:
      return {
        ...state,
        password: action.payload,
        passwordError: false,
      };
    case ACTION.FAILED_EMAIL:
      return {
        ...state,
        emailError: true,
      };
    case ACTION.FAILED_PASSWORD:
      return {
        ...state,
        passwordError: true,
      };
    case ACTION.LOADING_LOGIN:
      return {
        ...state,
        loadingLogin: action.payload,
      };
    default:
      return state;
  }
};

export type TUser = {
  displayName: string | null;
  email: string;
  emailVerified: boolean;
  isAnonymous: boolean;
  metadata: { creationTime: number; lastSignInTime: number };
  phoneNumber: string | null;
  photoURL: string | null;
  providerId: string;
  tenantId: null;
  uid: string;
};
