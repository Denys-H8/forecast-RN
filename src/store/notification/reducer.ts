import { FirebaseMessagingTypes } from '@react-native-firebase/messaging';
import * as ACTION from './actionTypes';

type Props = {
  routeName: string;
  notification: FirebaseMessagingTypes.RemoteMessage | null;
};

export const initialState: Props = {
  routeName: '',
  notification: null,
};

export const notificationReducer = (state = initialState, action: { type: string; payload?: any }) => {
  switch (action.type) {
    case ACTION.ADD_NOTIFICATION: {
      return {
        ...state,
        notification: action.payload,
      };
    }
    case ACTION.REMOVE_NOTIFICATION: {
      return {
        ...state,
        notification: null,
      };
    }

    default:
      return state;
  }
};
