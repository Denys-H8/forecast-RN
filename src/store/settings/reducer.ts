import * as ACTION from './actionTypes';

type TState = {
  name: string;
  age: string | number;
  dayPeriod: number;
  updateTime: number;
  temperatureScale: string;
};

export const initialState: TState = {
  name: '',
  age: '',
  dayPeriod: 5,
  updateTime: 10,
  temperatureScale: 'metric',
};

export const settingsReducer = (state = initialState, action: { type: string; payload?: any }) => {
  switch (action.type) {
    case ACTION.CHANGED_NAME:
      return {
        ...state,
        name: action.payload,
      };
    case ACTION.CHANGED_AGE:
      return {
        ...state,
        age: action.payload,
      };
    case ACTION.CHANGED_DAY_PERIOD:
      return {
        ...state,
        dayPeriod: action.payload,
      };
    case ACTION.CHANGED_UPDATE_TIME:
      return {
        ...state,
        updateTime: action.payload,
      };
    case ACTION.SET_TEMPERATURE_SCALE:
      return {
        ...state,
        temperatureScale: action.payload,
      };
    default:
      return state;
  }
};
