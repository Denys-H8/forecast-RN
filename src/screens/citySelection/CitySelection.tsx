import React, { useCallback, useEffect } from 'react';
import { FlatList, View, RefreshControl } from 'react-native';
import CityCard from './components/cityCard';
import SearchField from './components/searchField';
import { cities } from './cityData';
import { styles } from './citySelection.styles';
import { useDispatch, useSelector } from 'react-redux';
import { cityActions } from 'src/store/rootActions';
import { debounce } from 'lodash';
import * as CITY_SELECTORS from '../../store/citySelection/selectors';
import { TData, TGroupWeatherItem, TGroupWeather, TCity } from 'src/store/citySelection/reducer';

const CitySelection: React.FC = () => {
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(cityActions.resetQuery());
    dispatch(cityActions.getWeatherForCities());
  }, [dispatch]);

  const query = useSelector(CITY_SELECTORS.query);
  const isLoading = useSelector(CITY_SELECTORS.isLoadingSelector);

  const handleSearch = (query: string) => {
    const data: TData[] = cities.filter((item: TCityBig) => {
      return item.name.toLowerCase().includes(query.toLowerCase());
    });
    dispatch(cityActions.queryChanged(query, data));
    handleSearchSaga();
  };

  const handleSearchSaga = useCallback(
    debounce(() => {
      dispatch(cityActions.setLoading(true));
      dispatch(cityActions.getWeatherForCities());
    }, 700),
    [],
  );

  const handleCityPress = (city: TCity) => {
    dispatch(cityActions.pressedCity(city));
  };

  const keyExtractor = (item: TGroupWeatherItem) => item.id.toString();
  const listHeaderComponent = <SearchField onChangeText={handleSearch} value={query} />;
  const itemSeparatorComponent = () => <View style={styles.separator} />;
  const renderItem = ({ item }: any) => <CityCard item={item} onPress={handleCityPress} />;
  const weatherData: TGroupWeather = useSelector(CITY_SELECTORS.weatherData);

  return (
    <FlatList
      style={styles.container}
      keyExtractor={keyExtractor}
      data={weatherData.list}
      ListHeaderComponent={listHeaderComponent}
      stickyHeaderIndices={[0]}
      ItemSeparatorComponent={itemSeparatorComponent}
      renderItem={renderItem}
      refreshControl={<RefreshControl refreshing={isLoading} />}
    />
  );
};

type TCityBig = {
  id: number;
  name: string;
  state: string;
  country: string;
  coord: {
    lon: number;
    lat: number;
  };
};

export default CitySelection;
