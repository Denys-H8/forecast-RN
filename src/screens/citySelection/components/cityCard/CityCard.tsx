import React from 'react';
import { View, TouchableOpacity } from 'react-native';
import Icon from 'react-native-vector-icons/Fontisto';
import CustomText from 'src/components/text';
import constants from 'src/constants';
import { getWeatherIconName } from 'src/helpers';
import { TCity, TGroupWeatherItem } from 'src/store/citySelection/reducer';
import { styles } from './cityCard.styles';

type Props = {
  item: TGroupWeatherItem;
  onPress: (city: TCity) => void;
};

const CityCard: React.FC<Props> = ({ item, onPress }) => {
  const city = { id: item.id, name: item.name, lat: item.coord.lat, lon: item.coord.lon };

  return (
    <TouchableOpacity style={styles.itemContainer} testID="CityCard.TouchableOpacity" onPress={() => onPress(city)}>
      <View style={styles.degreesContainer}>
        <CustomText style={styles.number}>{item.main.temp.toFixed(1)}</CustomText>
        <CustomText style={styles.celsius}>0</CustomText>
      </View>
      <View style={styles.cityWeatherContainer}>
        <View style={styles.iconContainer}>
          <Icon
            name={getWeatherIconName(item.weather[0].description)}
            size={35}
            color={constants.SUB_COLOR}
            style={styles.icon}
          />
        </View>
        <View style={styles.cityContainer}>
          <CustomText style={styles.cityName}>{item.name}</CustomText>
          <CustomText style={styles.weatherName}>{item.weather[0].description.toUpperCase()}</CustomText>
        </View>
      </View>
    </TouchableOpacity>
  );
};

export default CityCard;
