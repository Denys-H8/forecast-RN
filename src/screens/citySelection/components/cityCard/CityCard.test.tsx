import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import CityCard from './CityCard';
import { TGroupWeatherItem } from 'src/store/citySelection/reducer';

export const itemMock: TGroupWeatherItem = {
  coord: {
    lon: 1525545,
    lat: 1024546,
  },
  sys: {
    sunrise: 1525545,
    sunset: 1024546,
  },
  weather: [
    {
      description: 'Sunny',
    },
  ],
  main: {
    temp: 20.7,
    feels_like: 23,
    temp_min: 15,
    temp_max: 21,
  },
  visibility: 95,
  wind: {
    speed: 10,
    deg: 10,
  },
  clouds: {
    all: 20,
  },
  dt: 1024546,
  id: 1024546,
  name: 'City',
};

test('CityCard', () => {
  const onPressMock = jest.fn();

  const { toJSON, getByTestId } = render(<CityCard item={itemMock} onPress={onPressMock} />);

  fireEvent.press(getByTestId('CityCard.TouchableOpacity'));

  expect(toJSON()).toMatchSnapshot();
  expect(onPressMock).toHaveBeenCalled();
});
