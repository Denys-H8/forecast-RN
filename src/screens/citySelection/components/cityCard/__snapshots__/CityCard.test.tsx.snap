// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CityCard 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "justifyContent": "space-between",
      "opacity": 1,
      "paddingLeft": 33,
      "paddingVertical": 20,
    }
  }
  testID="CityCard.TouchableOpacity"
>
  <View
    style={
      Object {
        "alignItems": "flex-start",
        "flexDirection": "row",
        "width": 80,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontFamily": "WorkSans-Regular",
          },
          Object {
            "color": "#fff",
            "fontSize": 36,
            "lineHeight": 42,
          },
        ]
      }
    >
      20.7
    </Text>
    <Text
      style={
        Array [
          Object {
            "fontFamily": "WorkSans-Regular",
          },
          Object {
            "color": "#fff",
            "fontSize": 18,
            "lineHeight": 18,
          },
        ]
      }
    >
      0
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "flex-start",
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "center",
          "width": 80,
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "#8B8EA5",
              "fontSize": 35,
            },
            Object {
              "marginRight": 28,
            },
            Object {
              "fontFamily": "Fontisto",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        î¨º
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "flex-start",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "WorkSans-Regular",
            },
            Object {
              "color": "#fff",
              "fontFamily": "WorkSans-SemiBold",
              "fontSize": 24,
              "lineHeight": 28,
            },
          ]
        }
      >
        City
      </Text>
      <Text
        style={
          Array [
            Object {
              "fontFamily": "WorkSans-Regular",
            },
            Object {
              "color": "#8B8EA5",
              "fontSize": 12,
              "lineHeight": 14,
            },
          ]
        }
      >
        SUNNY
      </Text>
    </View>
  </View>
</View>
`;
